### PR TITLE
dylink: Synchronize dlsym (as well as dlopen) calls between threads

### PR DESCRIPTION
**Pull request by sbc100**
Prior to this change we were tracking the sequence of `dlopen` events       
and replaying them on each thread during `dlsync`.                       
                                                                         
Now we also track `dlsym` events.  A new data structure is used here to  
abstract over these two events (`dlevent`).  During dlsync we then reply 
both the `dlopen` and the `dlsym` events.  For `dlsym` events we         
serialized them as "dso+index", i.e. which dso did the symbol come from  
and what is the index of the symbol within that dso's symbol table.   

This is important for #18376.

### CODE REVIEW

File: system/lib/libc/dynlink.c [hunk 1]
```diff
@@ -188,7 +231,7 @@ static void dlopen_js_onsuccess(struct dso* dso, struct async_data* data) {
       dso->mem_addr,
       dso->mem_size);
   load_library_done(dso);
-  do_write_lock();
+  do_write_unlock();
```
Comment (kripken) [inline system/lib/libc/dynlink.c:234]: This line made me sit back in my chair...

Is it a bugfix? Or how does it relate to this PR?

File: system/lib/libc/dynlink.c [hunk 2]
```diff
@@ -223,23 +266,25 @@ static void ensure_init() {
 void* dlopen(const char* file, int flags) {
   ensure_init();
   if (!file) {
-    return head;
+    return head->dso;
```
Comment (kripken) [inline system/lib/libc/dynlink.c:269]: What if this is a dlsym event and not another dlopen?

File: system/lib/libc/dynlink.c [hunk 3]
```diff
@@ -223,23 +266,25 @@ static void ensure_init() {
...
   dlsync();
 #endif
 
+  struct dso* p;
+
   /* Search for the name to see if it's already loaded */
-  for (p = head; p; p = p->next) {
-    if (!strcmp(p->name, file)) {
-      dbg("dlopen: already opened: %p", p);
+  for (struct dlevent* e = head; e; e = e->next) {
```
Comment (kripken) [inline system/lib/libc/dynlink.c:284]: Shouldn't we skip dlsym events (if only for efficiency)?

File: src/library_dylink.js [hunk 4]
```diff
@@ -36,7 +36,10 @@ var LibraryDylink = {
       // First look for the orig$ symbol which is the symbols without
       // any legalization performed.
       sym = wasmImports['orig$' + symName];
-      if (sym) return sym;
+      if (sym) {
+        symName = 'orig$' + symName;
+        return [sym, symName];
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/library_dylink.js:41]: (avoids `[0]`, `[1]` stuff)

File: system/lib/libc/dynlink.c (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/dynlink.c:234]: Yeah, this just a typo from a previous change.   It seems we don't have any test coverage for async dlopen + threads yet.  Adding those can happen as a separate change I think.

File: system/lib/libc/dynlink.c (hunk 2 above)
Comment (sbc100) [reply system/lib/libc/dynlink.c:269]: `head` is always the primary `__main__` dso.  Its ensured by `ensure_init`

File: system/lib/libc/dynlink.c (hunk 3 above)
Comment (sbc100) [reply system/lib/libc/dynlink.c:284]: Done.

**Review by kripken — APPROVED** [review]

**Pull request merged**